### PR TITLE
All the code

### PR DESCRIPTION
**Pull request by rodsouza9**
Review for all the code

### CODE REVIEW

File: src/App.css [hunk 1]
```diff
@@ -36,3 +36,114 @@
...
+    width : 10.78%;
+    height: 10.78%;
+    border-style: solid;
+    border-width: 1px;
+    border-color: blue;
+    float: left;
+    display: flex;
+    flex-wrap: wrap;
+    justify-content: center;
+    align-items: center;
```
Comment (advait) [inline src/App.css:50]: Get a linter that automatically sorts css properties in alpha order.

File: src/App.tsx [hunk 2]
```diff
@@ -1,26 +1,266 @@
...
     </div>
   );
+};
+
+type SquareAddress = number;
+
+interface BoardState {
+    highlights: SquareAddress[];
+    contradicts: SquareAddress[];
+    permanentValues: SquareAddress[];
```
Comment (advait) [inline src/App.tsx]: These might be better represented as `Set`s: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

File: src/App.tsx [hunk 3]
```diff
@@ -1,26 +1,266 @@
...
+interface BoardProps {
+    values: Array<SquareValue | null>;
+}
+
+class Board extends React.Component<BoardProps, BoardState> {
+    public state: BoardState = {
+        highlights: [37],
+        contradicts: [38],
+        permanentValues: [],
+        values: this.props.values,
```
Comment (advait) [inline src/App.tsx]: Hard rule: Do not ever set state values based on prop values.

Remove the `values` key.

File: src/App.tsx [hunk 4]
```diff
@@ -1,26 +1,266 @@
...
+      */
+    public toggleSelectSquare(i: SquareAddress) {
+        this.setState( {
+            ...this.state,
+            highlights: [i],
+            },
+        );
+    }
+
+    public getPermanentValues(): SquareAddress[] {
```
Comment (advait) [inline src/App.tsx]: What is this function for?

File: src/App.tsx [hunk 5]
```diff
@@ -1,26 +1,266 @@
...
+                list.push(null);
+            } else {
+                list.push(this.props.values[i] as SquareValue);
+            }
+        }
+        return list;
+    } */
+
+    public render() {
+        this.state.permanentValues = this.getPermanentValues();
```
Comment (advait) [inline src/App.tsx]: Rule: Don't ever set state via `this.state.foo = bar`. Always call `setState`.

Rule: Don't ever call `setState` in `render`

File: src/App.tsx [hunk 6]
```diff
@@ -1,26 +1,266 @@
...
+                  * click delete markings
+                  */
+            }
+        }
+        this.setState({
+            values: newValues,
+        });
+    }
+
+/*     convertUndefinedValues(): (SquareValue | null)[] {
```
Comment (advait) [inline src/App.tsx]: Commented code should not be committed. Delete it and refer to the commit history to retrieve it.

File: src/App.tsx [hunk 7]
```diff
@@ -1,26 +1,266 @@
...
+            if (!this.state.permanentValues.includes(address)) {
+                newValues[address] = null;
+                /**
+                  * //TODO: if marking exist on first delete click
+                  * delete only value and show markings on second
+                  * click delete markings
+                  */
+            }
+        }
+        this.setState({
```
Comment (advait) [inline src/App.tsx:91]: You are blowing away other parts of the state.

Add a `...this.state` before `values`

File: src/App.tsx [hunk 8]
```diff
@@ -1,26 +1,266 @@
...
+                const newValues = JSON.parse(JSON.stringify(this.state.values));
+                newValues[addy] = i;
+                this.setState({
+                    values: newValues,
+                });
+            }
+        }
+    }
+
+    public deleteSelectedSquares(): void {
```
Comment (advait) [inline src/App.tsx:79]: Please add a jsdoc comment explaining what this function is supposed to do.

File: src/App.tsx [hunk 9]
```diff
@@ -1,26 +1,266 @@
...
+            list.push(this.renderSquare(i));
+        }
+        return list;
+    }
+
+    public renderSquare(i: SquareAddress) {
+        const isHighlighted = this.state.highlights.includes(i);
+        const isContradicting = this.state.contradicts.includes(i);
+        const marks = i % 9 + 1 == 7 ? [i % 1 as SquareValue, i % 5 + 1 as SquareValue, i % 3 + 1 as SquareValue, i % 8 + 1 as SquareValue] : [];
+        return <Square value={this.state.values[i]}
```
Comment (advait) [inline src/App.tsx]: This line is too long. Please add a newline for each prop.

File: src/App.tsx [hunk 10]
```diff
@@ -1,26 +1,266 @@
...
+    isContradicting: boolean;
+    value: SquareValue | null;
+    markings: SquareValue[];
+    onClick: () => void;
+}
+
+type SquareValue = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 ;
+
+class Square extends React.Component<SquareProps, {}> {
+    public getMarks(): boolean[] {
```
Comment (advait) [inline src/App.tsx]: Please rewrite this as:

```js
public isMarked(squareValue: SquareValue): boolean
```

File: src/App.tsx [hunk 11]
```diff
@@ -1,26 +1,266 @@
...
+    public render() {
+        const light: string = this.props.isHighlighted ? "highlight" : (this.props.isContradicting ? "contradict" : "");
+        const marks: boolean[] = this.getMarks();
+        const displayMarkings: (boolean | 0) = // determine if markings or value should be displayed
+            this.props.markings && this.props.markings.length &&
+            !this.props.isPermanent && !this.props.value;
+        if (displayMarkings) {
+            return(
+                <div className={"square " + light} onClick={this.props.onClick}>
+                    <div className="mark">{marks[1] ? 1 : null}</div>
```
Comment (advait) [inline src/App.tsx]: Replace these with a loop

File: src/App.tsx [hunk 12]
```diff
@@ -1,26 +1,266 @@
...
+        const lst: boolean[] = [];
+        for (let i = 0; i < 9; i++) {
+            lst[i] = this.props.markings.includes(i as SquareValue);
+        }
+        return lst;
+    }
+    public render() {
+        const light: string = this.props.isHighlighted ? "highlight" : (this.props.isContradicting ? "contradict" : "");
+        const marks: boolean[] = this.getMarks();
+        const displayMarkings: (boolean | 0) = // determine if markings or value should be displayed
```
Comment (advait) [inline src/App.tsx]: `boolean | 0` is an inappropriate type. Please use `boolean`.

File: src/App.tsx [hunk 13]
```diff
@@ -1,26 +1,266 @@
...
+            this.props.markings && this.props.markings.length &&
```
Comment (advait) [inline src/App.tsx]: Please add a newline for each `&&`

File: src/App.tsx [hunk 14]
```diff
@@ -1,26 +1,266 @@
...
+        return (
+            <div className="button-num-pad">
+                {this.renderNumButtons()}
+                <Button onClick={this.props.onClickDel} className="button" variant="contained">DELETE</Button>
+            </div>
+        );
+    }
+}
+
+class NormalButton extends React.Component<{}, {}> {
```
Comment (advait) [inline src/App.tsx]: Please inline these four classes.

Inline means don't create separate classes for each of these.

File: src/App.tsx [hunk 15]
```diff
@@ -1,26 +1,266 @@
...
+            if (this.props.values[i]) {
+                list.push(i as SquareAddress);
+            }
+        }
+        return list;
+    }
+
+    public normalMarkSelectedSquares(i: SquareValue) {
+        for (let j = 0; j < this.state.highlights.length; j++) {
+            const addy: SquareAddress = this.state.highlights[j];
```
Comment (advait) [inline src/App.tsx]: `address`

File: src/App.tsx [hunk 16]
```diff
@@ -1,26 +1,266 @@
...
+            if (!this.state.permanentValues.includes(addy)) {
+                const newValues = JSON.parse(JSON.stringify(this.state.values));
+                newValues[addy] = i;
+                this.setState({
```
Comment (advait) [inline src/App.tsx]: `...this.state`.

**Review by advait — CHANGES_REQUESTED** [review]

File: src/App.tsx (hunk 3 above)
Comment (rodsouza9) [reply src/App.tsx]: I need a value key that represents all the values in reach square in the board component. But the permanent values are going to be passed in by the prop. So should I just initially set values to [] and then create a function called setValues() which reads the prop and copies the permanent values into the correct positions? and then call that function in the render() of the board. This is how I did the permanentValues key

File: src/App.tsx (hunk 11 above)
Comment (advait) [reply src/App.tsx]: Use the `.map` method on Arrays:
https://alligator.io/react/rendering-arrays-in-react/

File: src/App.tsx (hunk 10 above)
Comment (advait) [reply src/App.tsx]: On second thought, maybe replace with a `renderMarking(squareValue: SquareValue)` method.

File: src/App.tsx [hunk 17]
```diff
@@ -1,26 +1,236 @@
...
+    }
+
+    public renderSquare(i: SquareAddress) {
+        const isHighlighted = this.state.highlights.has(i);
+        const isContradicting = this.state.contradicts.has(i);
+        const marks = i % 9 + 1 == 7 ? [i % 1 as SquareValue, i % 5 + 1 as SquareValue, i % 3 + 1 as SquareValue, i % 8 + 1 as SquareValue] : [];
+        return <Square
+            value={this.isPermanent(i) ? this.props.permanentValues[i] : this.state.values[i]}
+            isPermanent={this.isPermanent(i)}
+            isHighlighted = {isHighlighted}
```
Comment (advait) [inline src/App.tsx]: Whitespace

File: src/App.tsx [hunk 18]
```diff
@@ -1,26 +1,236 @@
...
+
+class Square extends React.Component<SquareProps, {}> {
+    public getMarks(): (SquareValue | null)[] {
+        const lst: (SquareValue | null)[] = [];
+        for (let i = 0; i < 9; i++) {
+            lst[i] = this.props.markings.includes((i+1) as SquareValue) ? ((i+1) as SquareValue) : null;
+        }
+        return lst;
+    }
+    public renderMarkings() {
```
Comment (advait) [inline src/App.tsx]: Collapse the functionality of getMarks and rederMarkings into a single function.

File: src/App.tsx [hunk 19]
```diff
@@ -1,26 +1,236 @@
...
+        const marks: (SquareValue | null)[] = this.getMarks();
+        const list = [];
+        for (let i = 0; i < 9; i++) {
+            list.push(<div className="mark">{marks[i]}</div>);
+        }
+        return list;
+    }
+    public render() {
```
Comment (advait) [inline src/App.tsx:169]: New line

File: src/App.tsx [hunk 20]
```diff
@@ -1,26 +1,236 @@
...
+        const light: string = this.props.isHighlighted ? "highlight" : (this.props.isContradicting ? "contradict" : "");
+        const displayMarkings: boolean = // determine if markings or value should be displayed
+            this.props.markings &&
```
Comment (advait) [inline src/App.tsx]: This check is unnecessary, no?

**Review by advait — CHANGES_REQUESTED** [review]
Code is looking a lot better! Minor comments.

**Pull request merged**